File: Assets/Bundles/Shared/Resources/TS/Item/HeldItems/ProjectileLauncher/ProjectileLauncherHeldItem.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Flamework = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Flamework
local ItemPlayMode = require("Shared/TS/Entity/Animation/InventoryEntityAnimator").ItemPlayMode
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local GetItemMeta = require("Shared/TS/Item/ItemDefinitions").GetItemMeta
local ProjectileUtil = require("Shared/TS/Projectile/ProjectileUtil").ProjectileUtil
local Mouse = require("Shared/TS/UserInput/init").Mouse
local Bin = require("Shared/TS/Util/Bin").Bin
local CanvasAPI = require("Shared/TS/Util/CanvasAPI").CanvasAPI
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local OnLateUpdate = require("Shared/TS/Util/Timer").OnLateUpdate
local HeldItem = require("Shared/TS/Item/HeldItems/HeldItem").HeldItem
local ProjectileLauncherHeldItem
do
	local super = HeldItem
	ProjectileLauncherHeldItem = setmetatable({}, {
		__tostring = function()
			return "ProjectileLauncherHeldItem"
		end,
		__index = super,
	})
	ProjectileLauncherHeldItem.__index = ProjectileLauncherHeldItem
	function ProjectileLauncherHeldItem.new(...)
		local self = setmetatable({}, ProjectileLauncherHeldItem)
		return self:constructor(...) or self
	end
	function ProjectileLauncherHeldItem:constructor(...)
		super.constructor(self, ...)
		self.chargeBin = Bin.new()
		self.currentlyCharging = false
		self.startHoldTimeSec = 0
		self.projectileTrajectoryRenderer = GameObject:Find("ProjectileTrajectoryRenderer"):GetComponent("ProjectileTrajectoryRenderer")
	end
	function ProjectileLauncherHeldItem:OnChargeStart()
		super.OnChargeStart(self)
		if not self.entity:IsLocalCharacter() then
			return nil
		end
		if RunUtil:IsServer() then
			return nil
		end
		if not self.meta.ProjectileLauncher then
			return nil
		end
		local ammoItemMeta = GetItemMeta(self.meta.ProjectileLauncher.ammoItemType)
		local ammoMeta = ammoItemMeta.Ammo
		if CanvasAPI:IsPointerOverUI() then
			return nil
		end
		if not self:HasRequiredAmmo() then
			return nil
		end
		self.currentlyCharging = true
		local _result = self.entity.anim
		if _result ~= nil then
			_result:PlayItemUse(0, ItemPlayMode.HOLD)
		end
		self.startHoldTimeSec = os.clock()
		local mouse = Mouse.new()
		local localEntityController = Flamework.resolveDependency("Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController")
		self.chargeBin:Add(OnLateUpdate:ConnectWithPriority(200, function(deltaTime)
			if self.currentlyCharging then
				local chargeSec = os.clock() - self.startHoldTimeSec
				local launchPos = ProjectileUtil:GetLaunchPosition(self.entity, localEntityController:IsFirstPerson())
				local launchData = self:GetLaunchData(self.entity, mouse, self.meta, chargeSec, launchPos)
				self.projectileTrajectoryRenderer:UpdateInfo(launchPos, launchData.velocity, 0, ammoMeta.gravity)
			end
		end))
		self.chargeBin:Add(function()
			mouse:Destroy()
		end)
	end
	function ProjectileLauncherHeldItem:HasRequiredAmmo()
		if not self.entity:IsAlive() or not (TS.instanceof(self.entity, CharacterEntity)) then
			return false
		end
		local inventory = self.entity:GetInventory()
		local _launcherItemMeta = inventory:GetHeldItem()
		if _launcherItemMeta ~= nil then
			_launcherItemMeta = _launcherItemMeta:GetMeta()
		end
		local launcherItemMeta = _launcherItemMeta
		local _nullableProjectileLauncherMeta = launcherItemMeta
		if _nullableProjectileLauncherMeta ~= nil then
			_nullableProjectileLauncherMeta = _nullableProjectileLauncherMeta.ProjectileLauncher
		end
		local nullableProjectileLauncherMeta = _nullableProjectileLauncherMeta
		if nullableProjectileLauncherMeta then
			local launcherMeta = nullableProjectileLauncherMeta
			if inventory:HasItemType(launcherMeta.ammoItemType) then
				return true
			end
		end
		return false
	end
	function ProjectileLauncherHeldItem:OnUseClient(useIndex)
		super.OnUseClient(self, useIndex)
		if not self.entity:IsLocalCharacter() then
			return nil
		end
		self.currentlyCharging = false
		self.projectileTrajectoryRenderer:SetDrawingEnabled(false)
		if CanvasAPI:IsPointerOverUI() then
			return nil
		end
		if not self:HasRequiredAmmo() then
			return nil
		end
		local chargeSec = os.clock() - self.startHoldTimeSec
		--[[
			* All checks passed. Time to FIRE!
		]]
		self.startHoldTimeSec = -1
		local mouse = Mouse.new()
		local launchPos = ProjectileUtil:GetLaunchPosition(self.entity, (Flamework.resolveDependency("Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController")):IsFirstPerson())
		local launchData = self:GetLaunchData(self.entity, mouse, self.meta, chargeSec, launchPos)
		self.entity:LaunchProjectile(self.meta.ProjectileLauncher.ammoItemType, launchData.velocity)
		local _result = self.entity.anim
		if _result ~= nil then
			_result:PlayItemUse(1)
		end
	end
	function ProjectileLauncherHeldItem:OnCallToActionEnd()
		super.OnCallToActionEnd(self)
		if not self.entity:IsLocalCharacter() then
			return nil
		end
		self.currentlyCharging = false
		self.chargeBin:Clean()
	end
	function ProjectileLauncherHeldItem:GetLaunchData(entity, mouse, launcherItemMeta, chargeSec, launchPos)
		local launcherMeta = launcherItemMeta.ProjectileLauncher
		local ammoItemMeta = GetItemMeta(launcherMeta.ammoItemType)
		local ammoMeta = ammoItemMeta.Ammo
		local aimVector = self:GetAimVector(mouse, launchPos, ammoMeta)
		local launchForceData = ProjectileUtil:GetLaunchForceData(launcherItemMeta, aimVector, chargeSec)
		return {
			direction = launchForceData.direction,
			velocity = launchForceData.initialVelocity,
		}
	end
	function ProjectileLauncherHeldItem:GetAimVector(mouse, launchPosition, ammoMeta)
		-- Note: We could probably get more advanced with this calculation but this constant works pretty well.
		-- Alternatively, we could calculate the "distance" of the path and choose X percent along that distance.
		local aimDistance = 100
		local ray = Camera.main:ScreenPointToRay(mouse:GetLocation())
		local _direction = ray.direction
		local _vector3 = Vector3.new(0, ammoMeta.yAxisAimAdjust, 0)
		local scaledDirection = (_direction + _vector3) * aimDistance
		local targetPoint = ray.origin + scaledDirection
		local _launchPosition = launchPosition
		local adjustedDir = targetPoint - _launchPosition
		return adjustedDir
	end
end
return {
	ProjectileLauncherHeldItem = ProjectileLauncherHeldItem,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Item/HeldItems/ProjectileLauncher/ProjectileLauncherHeldItem.ts
import { Dependency } from "@easy-games/flamework-core";
import { LocalEntityController } from "Client/Controllers/Global/Character/LocalEntityController";
import { Crosshair } from "CoreShared/Crosshair/Crosshair";
import { ItemPlayMode } from "Shared/Entity/Animation/InventoryEntityAnimator";
import { CharacterEntity } from "Shared/Entity/Character/CharacterEntity";
import { Entity } from "Shared/Entity/Entity";
import { AmmoMeta, ItemMeta } from "Shared/Item/ItemMeta";
import { ProjectileUtil } from "Shared/Projectile/ProjectileUtil";
import { Mouse } from "Shared/UserInput";
import { Bin } from "Shared/Util/Bin";
import { CanvasAPI } from "Shared/Util/CanvasAPI";
import { RunUtil } from "Shared/Util/RunUtil";
import { SignalPriority } from "Shared/Util/Signal";
import { OnLateUpdate } from "Shared/Util/Timer";
import { HeldItem } from "../HeldItem";
import { ItemUtil } from "../../ItemUtil";

export class ProjectileLauncherHeldItem extends HeldItem {
	private chargeBin = new Bin();
	private currentlyCharging = false;
	private startHoldTimeSec = 0;
	private projectileTrajectoryRenderer =
		GameObject.Find("ProjectileTrajectoryRenderer").GetComponent<ProjectileTrajectoryRenderer>();

	protected override OnChargeStart(): void {
		super.OnChargeStart();
		if (!this.meta.ProjectileLauncher) return;

		//Play the items animation  (bow draw)
		this.PlayItemAnimation(0, true);

		if (RunUtil.IsClient()) {
			if (!this.entity.IsLocalCharacter()) return;

			const ammoItemMeta = ItemUtil.GetItemMeta(this.meta.ProjectileLauncher.ammoItemType);
			const ammoMeta = ammoItemMeta.Ammo!;

			if (CanvasAPI.IsPointerOverUI()) return;

			if (!this.HasRequiredAmmo()) return;

			this.chargeBin.Add(Crosshair.AddDisabler());

			this.currentlyCharging = true;

			this.entity.anim?.PlayItemUse(0, ItemPlayMode.HOLD);

			this.startHoldTimeSec = os.clock();

			const mouse = new Mouse();
			const localEntityController = Dependency<LocalEntityController>();
			this.chargeBin.Add(
				OnLateUpdate.ConnectWithPriority(SignalPriority.NORMAL, (deltaTime) => {
					if (this.currentlyCharging) {
						const chargeSec = os.clock() - this.startHoldTimeSec;

						const launchPos = ProjectileUtil.GetLaunchPosition(
							this.currentItemGOs,
							this.entity,
							localEntityController.IsFirstPerson(),
						);
						const launchData = this.GetLaunchData(this.entity, mouse, this.meta, chargeSec, launchPos);

						this.projectileTrajectoryRenderer.UpdateInfo(
							launchPos,
							launchData.velocity,
							0,
							ammoMeta.gravity,
						);
					}
				}),
			);
			this.chargeBin.Add(() => {
				mouse.Destroy();
			});
		}

		this.chargeBin.Add(
			this.entity.OnAdjustMove.Connect((moveModifier) => {
				moveModifier.blockSprint = true;
				moveModifier.speedMultiplier *= 0.4;
			}),
		);
	}

	private HasRequiredAmmo(): boolean {
		if (!this.entity.IsAlive() || !(this.entity instanceof CharacterEntity)) {
			return false;
		}

		const inventory = this.entity.GetInventory();
		const launcherItemMeta = inventory.GetHeldItem()?.GetMeta();
		const nullableProjectileLauncherMeta = launcherItemMeta?.ProjectileLauncher;

		if (nullableProjectileLauncherMeta) {
			const launcherMeta = nullableProjectileLauncherMeta!;

			if (inventory.HasItemType(launcherMeta.ammoItemType)) {
				return true;
			}
		}

		return false;
	}

	protected override TryChargeUse() {
		if (super.TryChargeUse()) {
			return true;
		} else {
			//Not charged up all the way
			this.entity.anim?.StartItemIdle();
			return false;
		}
	}

	protected override OnUseClient(useIndex: number): void {
		super.OnUseClient(useIndex);
		print("On use: " + useIndex);
		if (!this.entity.IsLocalCharacter()) return;

		this.currentlyCharging = false;

		if (CanvasAPI.IsPointerOverUI()) {
			return;
		}

		if (!this.HasRequiredAmmo()) {
			return;
		}

		const chargeSec = os.clock() - this.startHoldTimeSec;

		/*
		 * All checks passed. Time to FIRE!
		 */
		this.startHoldTimeSec = -1;

		const mouse = new Mouse();
		const launchPos = ProjectileUtil.GetLaunchPosition(
			this.currentItemGOs,
			this.entity,
			Dependency<LocalEntityController>().IsFirstPerson(),
		);
		const launchData = this.GetLaunchData(this.entity, mouse, this.meta, chargeSec, launchPos);
		this.entity.LaunchProjectile(
			this.meta.ProjectileLauncher!.ammoItemType,
			launchData.launchPos,
			launchData.velocity,
		);

		//Play the items animation  (bow shoot)
		this.PlayItemAnimation(1, false);
	}

	public override OnCallToActionEnd(): void {
		super.OnCallToActionEnd();
		this.currentlyCharging = false;
		this.chargeBin.Clean();
		this.projectileTrajectoryRenderer.SetDrawingEnabled(false);
	}

	public override OnUnEquip(): void {
		super.OnUnEquip();
		this.currentlyCharging = false;
		this.chargeBin.Clean();
		this.projectileTrajectoryRenderer.SetDrawingEnabled(false);
	}

	private GetLaunchData(
		entity: Entity,
		mouse: Mouse,
		launcherItemMeta: ItemMeta,
		chargeSec: number,
		launchPos: Vector3,
	): {
		direction: Vector3;
		launchPos: Vector3;
		velocity: Vector3;
	} {
		const launcherMeta = launcherItemMeta.ProjectileLauncher!;
		const ammoItemMeta = ItemUtil.GetItemMeta(launcherMeta.ammoItemType);
		const ammoMeta = ammoItemMeta.Ammo!;

		const aimVector = this.GetAimVector(mouse, launchPos, ammoMeta);

		const launchForceData = ProjectileUtil.GetLaunchForceData(launcherItemMeta, aimVector, chargeSec);

		return {
			direction: launchForceData.direction,
			launchPos: launchPos,
			velocity: launchForceData.initialVelocity,
		};
	}

	private GetAimVector(mouse: Mouse, launchPosition: Vector3, ammoMeta: AmmoMeta) {
		// Note: We could probably get more advanced with this calculation but this constant works pretty well.
		// Alternatively, we could calculate the "distance" of the path and choose X percent along that distance.
		const aimDistance = 100;

		const ray = Camera.main.ScreenPointToRay(mouse.GetLocation());
		const scaledDirection = ray.direction.add(new Vector3(0, ammoMeta.yAxisAimAdjust, 0)).mul(aimDistance);
		const targetPoint = ray.origin.add(scaledDirection);

		const adjustedDir = targetPoint.sub(launchPosition);

		return adjustedDir;
	}
}
